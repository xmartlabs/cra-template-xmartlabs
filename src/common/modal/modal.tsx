import { useEffect } from 'react';
import ReactDOM from 'react-dom';
import { classnames } from 'helpers/utils';
import { ReactComponent as CloseSVG } from '../../assets/icons/close.svg';
import modalStyles from './modal.module.scss';

export enum ModalSizes {
  small = 'small',
  medium = 'medium',
  big = 'big',
  unresticted = 'unresticted',
}

export interface ModalProps extends React.HTMLAttributes<HTMLDivElement> {
  size: ModalSizes
  isOpen: boolean;
  onClose: () => void;
  hideCloseButton?: boolean;
  children?: React.ReactNode;
  className?: string;
}

export const Modal = ({
  size, hideCloseButton = false, children, isOpen, onClose, className, ...props
}: ModalProps) => {
  useEffect(() => {
    const escFunction = (event: KeyboardEvent) => {
      if (event.key === 'Escape') {
        onClose();
      }
    };
    document.addEventListener('keydown', escFunction);

    return () => {
      document.removeEventListener('keydown', escFunction);
    };
  }, []);

  if (!isOpen) return null;

  return ReactDOM.createPortal(
    <div className={modalStyles.overlay}>
      <div
        role="dialog"
        aria-modal="true"
        className={classnames(modalStyles.modal, modalStyles[`modal-${size}`], className || '')}
        {...props}
      >
        {!hideCloseButton && (
          <button className={modalStyles.closeButton} type="button" onClick={onClose}>
            <CloseSVG aria-label="Close" />
          </button>
        )}
        {children}
      </div>
    </div>,
    document.getElementById('modal')!,
  );
};
