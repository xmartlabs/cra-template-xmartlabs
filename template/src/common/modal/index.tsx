export { Modal, ModalSizes } from './modal';
