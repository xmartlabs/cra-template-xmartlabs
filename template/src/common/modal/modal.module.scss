@use "sass:list";

.overlay {
  background-color: rgba(0, 0, 0, 0.3);
  position: fixed;
  width: 100vw;
  height: 100vh;
  top:0;
  right: 0;
  bottom: 0;
  left: 0;
  display: flex;
  cursor: default;
}

$modal-sizes: (
  small: (400px),
  medium: (600px),
  big: (968px),
);

.modal {
  /* Auto layout */
  display: flex;
  justify-content: center;
  padding: 24px;
  gap: 16px;

  /* Surface / Light / 10 */
  background: #FFFFFF;
  border-radius: 8px;
  border: solid black 1px;
  width: 100%;
  position: fixed;
  box-shadow: 0px 0px 18px 0px rgba(0, 0, 0, 0.75);
  transform: translate(-50%, -50%);
  top: 40%;
  left: 50%;

  // Sizes
  @each $size, $properties in $modal-sizes {
    &-#{$size} {
      max-width: list.nth($properties, 1)
    }
  }

  .closeButton {
    position: absolute;
    display: flex;
    margin: 12px;
    top: 0;
    right: 0;
    font-size: 20px;
    color: #8A8AA8;
    font-size: 24px;
  }

}
