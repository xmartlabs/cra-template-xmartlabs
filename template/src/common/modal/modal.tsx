import { useEffect } from 'react';
import ReactDOM from 'react-dom';
import { classnames } from 'helpers/utils';
import xSvg from '../../assets/icons/x.svg';
import modalStyles from './modal.module.scss';

export enum ModalSizes {
  small = 'small',
  medium = 'medium',
  big = 'big',
  unresticted = 'unresticted',
}

interface ModalProps extends React.HTMLAttributes<HTMLDivElement> {
  size: ModalSizes
  isOpen: boolean;
  onClose: () => void;
  hideCloseButton?: boolean;
  children?: React.ReactNode;
  className?: string;
}

export const Modal = ({
  size, hideCloseButton = false, children, isOpen, onClose, className, ...props
}: ModalProps) => {
  useEffect(() => {
    const escFunction = (event: KeyboardEvent) => {
      if (event.key === 'Escape') {
        onClose();
      }
    };
    document.addEventListener('keydown', escFunction);

    return () => {
      document.removeEventListener('keydown', escFunction);
    };
  }, []);

  if (!isOpen) return null;

  return ReactDOM.createPortal(
    <div className={modalStyles.overlay}>
      <div
        role="dialog"
        aria-modal="true"
        className={classnames(modalStyles.modal, modalStyles[`modal-${size}`], className || '')}
        {...props}
      >
        {!hideCloseButton && <button className={modalStyles.closeButton} type="button" onClick={onClose}><img alt="x icon" src={xSvg} /></button>}
        {children}
      </div>
    </div>,
    document.getElementById('modal')!,
  );
};
