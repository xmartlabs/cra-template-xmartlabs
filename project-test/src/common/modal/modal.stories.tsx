import { ComponentStory, ComponentMeta } from '@storybook/react';
import styles from 'pages/home/home.module.scss';
import { Button } from 'common/button';
import { useState } from 'react';
import { Modal, ModalSizes } from './modal';

export default {
  title: 'Common/Modal',
  component: Modal,
  argTypes: {
    size: {
      control: {
        type: 'select',
      },
    },
  },
  args: {
    size: ModalSizes.medium,
    hideCloseButton: false,
    isOpen: true,
    children:
  <span>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit,
    sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam,
    quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.
    Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat
    nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia
    deserunt mollit anim id est laborum.
  </span>,
  },
} as unknown as ComponentMeta<typeof Modal>;

const Template: ComponentStory<typeof Modal> = (args) => {
  const [isOpen, setIsOpen] = useState(true);
  return isOpen ? <Modal {...args} onClose={() => setIsOpen(false)} /> : <div />;
};

export const Default = Template.bind({});

export const WithSizeBig = Template.bind({});
WithSizeBig.args = {
  size: ModalSizes.big,
};

export const WithSizeSmall = Template.bind({});
WithSizeSmall.args = {
  size: ModalSizes.small,
};

export const WithoutCloseButton = Template.bind({});
WithoutCloseButton.args = {
  hideCloseButton: true,
};

export const WithChildren = Template.bind({});
WithChildren.args = {
  children:
  <div className={styles.modalContent}>
    <h3 className={styles.modalTitle}>Title</h3>
    <span>
      Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac
      turpis egestas. Vestibulum tortor quam, feugiat vitae, ultricies eget, tempor
      sit amet, ante. Donec eu libero sit amet quam egestas semper. Aenean ultricies
      mi vitae est. Mauris placerat eleifend leo.
    </span>
    <div className={styles.modalFooter}>
      <Button bColor="secondary">Cancel</Button>
      <Button>Accept</Button>
    </div>
  </div>
  ,
};
