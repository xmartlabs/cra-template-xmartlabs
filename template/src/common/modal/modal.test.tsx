import { render } from '@testing-library/react';
import { Modal, ModalSizes } from './modal';

describe('Modal', () => {
  it('should render successfully', async () => {
    let portalRoot = document.getElementById('modal');
    if (!portalRoot) {
      portalRoot = document.createElement('div');
      portalRoot.setAttribute('id', 'modal');
      document.body.appendChild(portalRoot);
    }
    const { baseElement } = render(<Modal isOpen size={ModalSizes.medium} onClose={() => true} />);
    expect(baseElement.firstChild).toMatchSnapshot();
  });

  it('should render successfully with correct size', async () => {
    let portalRoot = document.getElementById('modal');
    if (!portalRoot) {
      portalRoot = document.createElement('div');
      portalRoot.setAttribute('id', 'modal');
      document.body.appendChild(portalRoot);
    }
    const { baseElement } = render(
      <Modal isOpen size={ModalSizes.medium} onClose={() => true}>
        <div>
          <h3>Title</h3>
          <span>
            Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac
            turpis egestas. Vestibulum tortor quam, feugiat vitae, ultricies eget, tempor
            sit amet, ante. Donec eu libero sit amet quam egestas semper. Aenean ultricies
            mi vitae est. Mauris placerat eleifend leo.
          </span>
        </div>
      </Modal>,
    );
    const modal = baseElement.querySelector('.modal');
    expect(modal).toHaveClass('modal-medium');
  });
});
