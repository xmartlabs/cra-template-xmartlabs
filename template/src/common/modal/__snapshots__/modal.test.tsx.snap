// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal should render successfully 1`] = `
<div
  id="modal"
>
  <div
    class="overlay"
  >
    <div
      aria-modal="true"
      class="modal modal-medium"
      role="dialog"
    >
      <button
        class="closeButton"
        type="button"
      >
        <svg
          aria-label="Close"
        >
          close.svg
        </svg>
      </button>
    </div>
  </div>
</div>
`;
